src/osd/ReplicatedPG.cc

```diff
@@ -2719,9 +2697,50 @@ void ReplicatedPG::promote_object(ObjectContextRef obc,
 
   assert(obc->is_blocked());
 
+  info.stats.stats.sum.num_promote++;
+}
+
+void ReplicatedPG::queue_async_promote(ObjectContextRef obc,
+                                       const hobject_t& missing_oid,
+                                       const object_locator_t& loc,
+                                       OpRequestRef op,
+                                       ObjectContextRef *promote_obc)
+{
+  hobject_t hoid = obc ? obc->obs.oi.soid : missing_oid;
+
+  assert(hoid != hobject_t());
+  if (scrubber.write_blocked_by_scrub(hoid, get_sort_bitwise())) {
+    dout(10) << __func__ << " " << hoid
+	     << " blocked by scrub" << dendl;
+    if (op) {
+      waiting_for_active.push_back(op);
+      dout(10) << __func__ << " " << hoid
+	       << " placing op in waiting_for_active" << dendl;
+    } else {
+      dout(10) << __func__ << " " << hoid
+	       << " no op, dropping on the floor" << dendl;
+    }
+    return;
+  }
+  if (!obc) { // we need to create an ObjectContext
```

Comment:
I think this isn't quite right.  If promote_obc != null, we need to unconditionally promote (please look at the _rollback_to() caller).  If it is null, then I don't think we need the obc even?  Please check the callers...